lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+    $file_content =~ s/$srch3/$rpt_data/;
+
+    my $report_file_path = catfile($self->{genome_dir}, 'genome_report.html');
+    my $fh2 = $self->_openWrite($report_file_path);
+    print $fh2 $file_content;
+    close $fh2;
+    #print $file_content;
+
+    my($vol, $f_path, $rfile) = splitpath($report_file_path);
+    my @html_report = ({'path'=> $report_file_path,
+                        'name'=> $rfile,
+                        'label'=> $rfile,
+                        'description'=> $report_title});
+    return @html_report;
+}
+
+#Create a KBaseReport with brief info/stats on a reannotated genome
+sub _generate_genome_report {
+    my ($self, $aa_ref, $aa_gff_conts, $func_tab, $ftr_cnt, $msg) = @_;
+
+    unless ($self->_validate_KB_objref($aa_ref)) {
+		return {"output_genome_ref"=>$aa_ref,
+                "workspace_name"=>undef,
+                "report_name"=>undef,
+                "report_ref"=>undef};
+    }
+
+    my $gn_info = $self->_fetch_object_info($aa_ref);
+    my $gn_ws = $gn_info->[7];
+
+    my %aa_stats = $self->_generate_stats_from_aa($aa_ref);
+    my %aa_gff_stats = $self->_generate_stats_from_gffContents($aa_gff_conts);
+    my %subsys_info = $self->_fetch_subsystem_info();
+
+    my @html_files = $self->_write_html_from_stats(\%aa_stats,
+                                                   \%aa_gff_stats,
+                                                   \%subsys_info,
+                                                   $func_tab);
+
+    my $kbr = new installed_clients::KBaseReportClient($self->{call_back_url});
+    my $report_message = $msg;
+    my $report_info = $kbr->create_extended_report(
+            {"message"=>$report_message,
+             "objects_created"=>[{"ref"=>$aa_ref, "description"=>"RAST re-annotated genome"}],
+             "html_links"=> \@html_files,
+             "direct_html_link_index"=> 0,
+             "html_window_height"=> 366,
+             "report_object_name"=>"kb_RAST_genome_report_".$self->_create_uuid(),
+             "workspace_name"=>$gn_info->[7]
+        });
+
+    return {"output_genome_ref"=>$aa_ref,
+            "workspace_name"=>$report_info->{workspace_name},
+            "report_name"=>$report_info->{name},
+            "report_ref"=>$report_info->{ref}};
+}
+
+#
+## Check the given ref string to make sure it conforms to the expected
+## object reference format and actually exists in KBase.
+## Return the object data if it passes, die otherwise.
+#
+sub _fetch_object_data {
+    my ($self, $obj_ref) = @_;
+    my $ret_obj_data = {};
+
+    unless ($obj_ref =~ m/^\d+\/\d+\/\d+$/) {
+		print "invalid object reference: $obj_ref\n";
+        return $ret_obj_data;
+    }
+
+    eval {
+        $ret_obj_data = $self->{ws_client}->get_objects2(
+                            {'objects'=>[{ref=>$obj_ref}]}
+                        )->{data}->[0]->{data};
+    };
+    if ($@) {
+        croak "ERROR Workspace.get_objects2 failed to access $obj_ref:$@\n";
+    }
```

Comment:
switch this out and use [Try::Tiny](https://metacpan.org/pod/Try::Tiny) instead.

Reply:
That's a lot.  I made all conversions in both AnnotationUtils.pm and MetagenomeUtils.pm.